### PR TITLE
W-14610032: extract needed logic from `mule-core`'s `RetryPolicyTemplate` to `mule-api`

### PR DESCRIPTION
**Pull request by rbourbonnavarro**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/api/retry/RetryCallback.java [hunk 1]
```diff
@@ -6,21 +6,19 @@
...
 
-import org.mule.runtime.core.api.retry.policy.RetryPolicy;
-
 /**
  * This is the main Retry SPI. The code inside the {@link #doWork} method is what will actually get <u>retried</u> according to
  * the {@link RetryPolicy} that has been configured. Note that retries can be wrapped in a transaction to ensure the work is
  * atomic.
  */
-public interface RetryCallback {
+@Deprecated
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/retry/RetryCallback.java:14]: add the javadoc indicating the replacement

File: core/src/main/java/org/mule/runtime/core/api/retry/RetryContext.java [hunk 2]
```diff
@@ -7,64 +7,12 @@
...
 import org.mule.api.annotation.NoImplement;
-import org.mule.runtime.api.message.Message;
-import org.mule.runtime.api.notification.NotificationDispatcher;
-
-import java.util.Map;
 
 /**
  * The RetryContext is used to store any data which carries over from attempt to attempt such as response messages.
  */
+@Deprecated
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/retry/RetryContext.java:14]: indicate the replacement in the javadoc

File: core/src/main/java/org/mule/runtime/core/api/retry/RetryNotifier.java [hunk 3]
```diff
@@ -12,12 +12,24 @@
  * This interface is a callback that allows actions to be performed after each retry attempt, such as firing notifications,
  * logging, etc.
  */
+@Deprecated
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/retry/RetryNotifier.java:15]: indicate the replacement in the javadoc

File: core/src/main/java/org/mule/runtime/core/api/retry/policy/PolicyStatus.java [hunk 4]
```diff
@@ -17,7 +17,7 @@
  * For example, a RetryPolicy may have a RetryCount - how many times the policy can be invoked. Once the retryCount has been
  * reached, the policy is exhausted and cannot be used again.
  */
-public final class PolicyStatus {
+public final class PolicyStatus implements org.mule.runtime.retry.api.policy.PolicyStatus {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/retry/policy/PolicyStatus.java:20]: deprecate?

File: core/src/main/java/org/mule/runtime/core/api/retry/policy/RetryPolicy.java [hunk 5]
```diff
@@ -6,113 +6,17 @@
...
-import java.util.function.Predicate;
-import java.util.function.Supplier;
-
-import org.reactivestreams.Publisher;
 
 /**
  * A RetryPolicy takes some action each time an exception occurs and returns a {@link PolicyStatus} which indicates whether the
  * policy is exhausted or should continue to retry.
  */
+@Deprecated
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/retry/policy/RetryPolicy.java:15]: indicate the replacement in the javadoc

File: core/src/main/java/org/mule/runtime/core/api/retry/policy/RetryPolicyTemplate.java [hunk 6]
```diff
@@ -32,18 +20,28 @@
  * A {@link RetryNotifier} may be set in order to take action upon each retry attempt.
  */
 @NoImplement
-public interface RetryPolicyTemplate {
+public interface RetryPolicyTemplate extends org.mule.runtime.retry.api.policy.RetryPolicyTemplate {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/retry/policy/RetryPolicyTemplate.java:23]: deprecate?

File: mule-retry-api/api-changes.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  "1.0.0": {
+    "revapi": {
+      "ignore": [
+        {
+          "ignore": true,
+          "code": "java.class.externalClassExposedInAPI",
+          "new": "interface org.reactivestreams.Publisher<T>",
+          "classSimpleName": "Publisher",
+          "package": "org.reactivestreams",
+          "classQualifiedName": "org.reactivestreams.Publisher",
```
Comment (elrodro83) [inline mule-retry-api/api-changes.json:11]: mule-api does not have this dependency

File: mule-retry-api/api-changes.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+          "elementKind": "interface",
+          "justification": "W-14610032: replace retry configuration API from `mule-core`"
+        },
+        {
+          "ignore": true,
+          "code": "java.class.externalClassExposedInAPI",
+          "new": "interface org.reactivestreams.Subscriber<T>",
+          "classSimpleName": "Subscriber",
+          "package": "org.reactivestreams",
+          "classQualifiedName": "org.reactivestreams.Subscriber",
```
Comment (elrodro83) [inline mule-retry-api/api-changes.json:21]: mule-api does not have this dependency

File: mule-retry-api/api-changes.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+          "elementKind": "interface",
+          "justification": "W-14610032: replace retry configuration API from `mule-core`"
+        },
+        {
+          "ignore": true,
+          "code": "java.class.externalClassExposedInAPI",
+          "new": "interface org.reactivestreams.Subscription",
+          "classSimpleName": "Subscription",
+          "package": "org.reactivestreams",
+          "classQualifiedName": "org.reactivestreams.Subscription",
```
Comment (elrodro83) [inline mule-retry-api/api-changes.json:31]: mule-api does not have this dependency

File: mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/RetryPolicy.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+   * @param shouldRetry   a predicate which evaluates each item to know if it should be retried or not
+   * @param onExhausted   an action to perform when the retry action has been exhausted
+   * @param errorFunction function used to map cause exception to exception emitted by retry policy.
+   * @param <T>           the generic type of the publisher's content
+   * @return a {@link Publisher} configured with the retry policy.
+   * @since 4.0
+   * @deprecated Use {@link #applyPolicy(Publisher, Predicate, Consumer, Function, Scheduler)} instead
+   */
+  @Deprecated
+  default <T> Publisher<T> applyPolicy(Publisher<T> publisher,
```
Comment (elrodro83) [inline mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/RetryPolicy.java:89]: this method uses reactivestreams, it should not be part of the api. Isn't this used only by the container?

File: mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/RetryPolicy.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+   * @param publisher      a publisher with the items which might fail
+   * @param shouldRetry    a predicate which evaluates each item to know if it should be retried or not
+   * @param onExhausted    an action to perform when the retry action has been exhausted
+   * @param errorFunction  function used to map cause exception to exception emitted by retry policy.
+   * @param retryScheduler the scheduler to use when retrying. If empty, an internal reactor Scheduler will be used.
+   * @param <T>            the generic type of the publisher's content
+   * @return a {@link Publisher} configured with the retry policy.
+   * @since 4.2
+   */
+  default <T> Publisher<T> applyPolicy(Publisher<T> publisher,
```
Comment (elrodro83) [inline mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/RetryPolicy.java:109]: this method uses reactivestreams, it should not be part of the api. Isn't this used only by the container?

File: mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/RetryPolicyTemplate.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+   * the retry logic.
+   *
+   * @param publisher a publisher with the items which might fail
+   * @param <T>       the generic type of the publisher's content
+   * @return a {@link Publisher} configured with the retry policy.
+   * @since 4.0
+   * @deprecated Use {@link #applyPolicy(Publisher, Scheduler)} instead
+   */
+  @Deprecated
+  default <T> Publisher<T> applyPolicy(Publisher<T> publisher) {
```
Comment (elrodro83) [inline mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/RetryPolicyTemplate.java:68]: this method uses reactivestreams, it should not be part of the api. Isn't this used only by the container?

File: mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/RetryPolicyTemplate.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+   * Applies the retry policy in a non blocking manner by transforming the given {@code publisher} into one configured to apply
+   * the retry logic.
+   *
+   * @param publisher      a publisher with the items which might fail
+   * @param retryScheduler the scheduler to use when retrying. If empty, an internal reactor Scheduler will be used.
+   * @param <T>            the generic type of the publisher's content
+   * @return a {@link Publisher} configured with the retry policy.
+   * @since 4.2
+   */
+  default <T> Publisher<T> applyPolicy(Publisher<T> publisher, Scheduler retryScheduler) {
```
Comment (elrodro83) [inline mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/RetryPolicyTemplate.java:92]: this method uses reactivestreams, it should not be part of the api. Isn't this used only by the container?

File: mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/RetryPolicyTemplate.java [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+   * @param publisher      a publisher with the items which might fail
+   * @param shouldRetry    a predicate which evaluates each item to know if it should be retried or not
+   * @param onExhausted    an action to perform when the retry action has been exhausted
+   * @param errorFunction  function used to map cause exception to exception emitted by retry policy.
+   * @param retryScheduler the scheduler to use when retrying. If empty, an internal reactor Scheduler will be used.
+   * @param <T>            the generic type of the publisher's content
+   * @return a {@link Publisher} configured with the retry policy.
+   * @since 4.2
+   */
+  default <T> Publisher<T> applyPolicy(Publisher<T> publisher,
```
Comment (elrodro83) [inline mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/RetryPolicyTemplate.java:131]: this method uses reactivestreams, it should not be part of the api. Isn't this used only by the container?

File: mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/util/rx/RxUtils.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2023 Salesforce, Inc. All rights reserved.
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.retry.api.policy.util.rx;
+
+public class RxUtils {
```
Comment (elrodro83) [inline mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/util/rx/RxUtils.java:9]: do you really need this here?

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Comment by rbourbonnavarro** [comment]
--validate

**Review by asanguinetti — APPROVED** [review]

**Pull request merged**